### ISSUE TITLE
🐞Need to check pointer for nil in `WithMountedFile`? User reported.

### ISSUE THREAD

**Issue by jpadams** [dagger/dagger#5696]
### What is the issue?

Encountered by @verdverm  here: https://discord.com/channels/707636530424053791/1144428151766790184/1144437805632716922

https://github.com/dagger/dagger-go-sdk/blob/v0.8.4/api.gen.go#L1027

### SDK version

Go SDK

### OS version

?

cc @vito @helderco

**Comment by verdverm** [comment]
it looks like many or all functions in that file do not validate their inputs

https://github.com/dagger/dagger-go-sdk/blob/v0.8.4/api.gen.go#L999

This kind of makes sense, given the pipeline style of the SDKs, so you cannot really return errors

I wonder if there is a possibility to have a second set of similar functions, that have the error checking and do not support pipelining. It would be really cool if you can call a bunch of piplined and end with an error based one

```go
func (r *Container) WithMountedFile(path string, source *File, opts ...ContainerWithMountedFileOpts) *Container {
  ...
}

func (r *Container) WithMountedFileE(path string, source *File, opts ...ContainerWithMountedFileOpts) (*Container, error) {
  if source == nil {
    return r, fmt.Errorf("...")
  }
  
  return C.WithMountedFile(path, source, opts...), nil
}

**Comment by TomChv** [comment]
> I wonder if there is a possibility to have a second set of similar functions, that have the error checking and do not support pipelining. It would be really cool if you can call a bunch of piplined and end with an error based one

Users could also check for null pointers before using Dagger, but I agree that it would be better if we check it inside dagger.

This idea would definitely make our API safer. wdyt @vito @shykes?
I can work on this week if you think it's a good idea?

**Comment by vito** [comment]
IMO this shouldn't really be our responsibility to guard against. You could cause this kind of panic with tons of Go APIs. Any error we raise would probably be less helpful than the panic, which will show you exactly what line caused it.

**Comment by TomChv** [comment]
> IMO this shouldn't really be our responsibility to guard against. You could cause this kind of panic with tons of Go APIs. Any error we raise would probably be less helpful than the panic, which will show you exactly what line caused it.

Thanks for your comment! Should we close this issue then?

**Comment by vito** [comment]
I think so - I'll go ahead and do that to not leave a loose end, but others can feel free to comment / vouch for reopening!

**Comment by verdverm** [comment]
The main problem here is the panic has no lines to our own code, so it was very difficult to link to.

It seems that a parallel set of SDK functions could be generated quite easily, that wraps the existing one but also validates input.

**Comment by vito** [comment]
@verdverm I'll reopen since I misunderstood this to mean `if self == nil` style checks, whereas this is actually about the objects passed as inputs to other methods. I can see how the backtrace would be confusing, because the `nil` panic doesn't happen much later until the lazy query is finally evaluated.

Even so I don't think having the increased API size is worth it considering developers are unlikely to know exactly when to run the version that validates inputs. Worst case they call it all the time and supremely clutter their codebase. But I think it's more likely that you just wouldn't think to use it because it's an unexpected state in the first place.

Maybe we could somehow annotate inputs with their original call site as they are added to the query? :thinking:

**Comment by wingyplus** [comment]
I'm not sure if we could do assert somehow and let crash in the API function since `nil` is unexpected input?

**Comment by wingyplus** [comment]
Like:

```go
func (...) WithMountedFile(path string, file *File, ...) *Container {
  AssertNotNil(file) // panic if it's nil.
  ....
}

**Comment by verdverm** [comment]
this seems like an SDK / lang specific thing, some languages might give better trace without this, but some languages like Go will need this for the DX

**Comment by TomChv** [comment]
> this seems like an SDK / lang specific thing, some languages might give better trace without this, but some languages like Go will need this for the DX

You are right, I first thought that we could handle it on server side but the GQL query is actually not even sent so I tried to trigger the error on our 3 SDK: Go, Ts and Python, (I excluded Rust and Elixir since I lack of experience in these language) and see what happens

## TL;DR

I'll update Go SDK, Python and Typescript look good enough to me for now

## Go

<details>
<summary>Code</summary>

```go
package main

import (
	"context"
	"fmt"
	"os"

	"dagger.io/dagger"
)

func main() {
	ctx := context.Background()

	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
	if err != nil {
		panic(err)
	}
	_, err = client.Container().WithMountedFile("/test", nil).Sync(ctx)

	fmt.Println(err)
}
```

</details>

<details>
<summary>Output</summary>

```shell
 │ panic: runtime error: invalid memory address or nil pointer dereference                                                                                                           
┃ │ [signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x105206550]                                                                                                        
┃ │                                                                                                                                                                                   
┃ │ goroutine 22 [running]:                                                                                                                                                           
┃ │ dagger.io/dagger.(*File).ID(0x1052e67c0?, {0x10531b820?, 0x1400008c1e0?})                                                                                                         
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/api.gen.go:1792 +0x20                                                                                                   
┃ │ dagger.io/dagger.(*File).XXX_GraphQLID(0x140000c1528?, {0x10531b820?, 0x1400008c1e0?})                                                                                            
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/api.gen.go:1815 +0x24                                                                                                   
┃ │ reflect.Value.call({0x1052fbc20?, 0x0?, 0x13?}, {0x105208d37, 0x4}, {0x140000c1cc0, 0x1, 0x0?})                                                                                   
┃ │         /Users/tomchauveau/go/go1.20/src/reflect/value.go:586 +0x838                                                                                                              
┃ │ reflect.Value.Call({0x1052fbc20?, 0x0?, 0x140000424d8?}, {0x140000c1cc0?, 0x10529a94e?, 0x10529a94c?})                                                                            
┃ │         /Users/tomchauveau/go/go1.20/src/reflect/value.go:370 +0x90                                                                                                               
┃ │ dagger.io/dagger/internal/querybuilder.marshalCustom({0x10531b820?, 0x1400008c1e0}, {0x1052fbc20?, 0x0?, 0x140000ac3c0?})                                                         
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/internal/querybuilder/marshal.go:138 +0x130                                                                             
┃ │ dagger.io/dagger/internal/querybuilder.marshalValue({0x10531b820, 0x1400008c1e0}, {0x1052fbc20?, 0x0?, 0x0?})                                                                     
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/internal/querybuilder/marshal.go:60 +0x5b4                                                                              
┃ │ dagger.io/dagger/internal/querybuilder.MarshalGQL({0x10531b820?, 0x1400008c1e0?}, {0x1052fbc20?, 0x0?})                                                                           
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/internal/querybuilder/marshal.go:53 +0xd0                                                                               
┃ │ dagger.io/dagger/internal/querybuilder.(*argument).marshal.func1()                                                                                                                
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/internal/querybuilder/querybuilder.go:179 +0x3c                                                                         
┃ │ sync.(*Once).doSlow(0x140000a0000?, 0x0?)                                                                                                                                         
┃ │         /Users/tomchauveau/go/go1.20/src/sync/once.go:74 +0x100                                                                                                                   
┃ │ sync.(*Once).Do(...)                                                                                                                                                              
┃ │         /Users/tomchauveau/go/go1.20/src/sync/once.go:65                                                                                                                          
┃ │ dagger.io/dagger/internal/querybuilder.(*argument).marshal(0x0?, {0x10531b820?, 0x1400008c1e0?})                                                                                  
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/internal/querybuilder/querybuilder.go:178 +0x6c                                                                         
┃ │ dagger.io/dagger/internal/querybuilder.(*Selection).marshalArguments.func1()                                                                                                      
┃ │         /Users/tomchauveau/Documents/DAGGER/dagger/sdk/go/internal/querybuilder/querybuilder.go:73 +0x28                                                                          
┃ │ golang.org/x/sync/errgroup.(*Group).Go.func1()                                                                                                                                    
┃ │         /Users/tomchauveau/go/pkg/mod/golang.org/x/sync@v0.3.0/errgroup/errgroup.go:75 +0x5c                                                                                      
┃ │ created by golang.org/x/sync/errgroup.(*Group).Go                                                                                                                                 
┃ │         /Users/tomchauveau/go/pkg/mod/golang.org/x/sync@v0.3.0/errgroup/errgroup.go:72 +0x9c                                                                                      
┃ │ exit status 2                                                                                                                                                                     
┻ │ 
  ┻ 
• Engine: c06b6d9d3801 (version main)

```

</details>

As you all mentioned above, the stacktrace is almost impossible to understand, specially because we cannot even get the line where the `nil` is inserted.

I'll update our SDK to assert that pointers are not `nil` on input and panic otherwise. It's still quite brutal to simply panic but it will be way better than what we currently have. We can then follow up with other improvement until we find the right way to handle nil input on Go SDK. 

## Python

<details>
<summary>Code</summary>

```python
"""Execute a command."""

import sys

import anyio

import dagger


async def test():
    async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
        _ = await client.container().with_mounted_file("/test", None).sync()


anyio.run(test)
```

</details>

<details>
<summary>Output</summary>

```shell
┃ │ Traceback (most recent call last):                                                                                                                                                
┃ │   File "/Users/tomchauveau/Documents/DAGGER/dagger/test.py", line 15, in <module>                                                                                                 
┃ │     anyio.run(test)                                                                                                                                                               
┃ │   File "/Users/tomchauveau/Documents/DAGGER/dagger/.venv/lib/python3.11/site-packages/anyio/_core/_eventloop.py", line 66, in run                                                 
┃ │     return async_backend.run(func, args, {}, backend_options)                                                                                                                     
┃ │            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                     
┃ │   File "/Users/tomchauveau/Documents/DAGGER/dagger/.venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 1968, in run                                             
┃ │     return native_run(wrapper(), debug=debug)                                                                                                                                     
┃ │            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                     
┃ │   File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run                                  
┃ │     return runner.run(main)                                                                                                                                                       
┃ │            ^^^^^^^^^^^^^^^^                                                                                                                                                       
┃ │   File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run                                  
┃ │     return self._loop.run_until_complete(task)                                                                                                                                    
┃ │            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                    
┃ │   File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete               
┃ │     return future.result()                                                                                                                                                        
┃ │            ^^^^^^^^^^^^^^^                                                                                                                                                        
┃ │   File "/Users/tomchauveau/Documents/DAGGER/dagger/.venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 1961, in wrapper                                         
┃ │     return await func(*args)                                                                                                                                                      
┃ │            ^^^^^^^^^^^^^^^^^                                                                                                                                                      
┃ │   File "/Users/tomchauveau/Documents/DAGGER/dagger/test.py", line 12, in test                                                                                                     
┃ │     _ = await client.container().with_mounted_file("/test", None).sync()                                                                                                          
┃ │               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                 
┃ │   File "/Users/tomchauveau/Documents/DAGGER/dagger/.venv/lib/python3.11/site-packages/dagger/client/_guards.py", line 137, in wrapper                                             
┃ │     with _handle_exception():                                                                                                                                                     
┃ │   File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 155, in __exit__                                  
┃ │     self.gen.throw(typ, value, traceback)                                                                                                                                         
┃ │   File "/Users/tomchauveau/Documents/DAGGER/dagger/.venv/lib/python3.11/site-packages/dagger/client/_guards.py", line 122, in _handle_exception                                   
┃ │     raise err from None                                                                                                                                                           
┃ │ TypeError: Method dagger.Container.with_mounted_file() parameter source="None" violates type hint 'File', as <class "builtins.NoneType"> "None" not instance of <protocol "dagger.
┃ │ le">.                                                                                                                                                                             
┻ │ 
  ┻ 
• Engine: c06b6d9d3801 (version main)
```

</details>

The stacktrace is big too but the last error message describes the problem explicitly.

```
┃ │ TypeError: Method dagger.Container.with_mounted_file() parameter source="None" violates type hint 'File', as <class "builtins.NoneType"> "None" not instance of <protocol "dagger.
```

I do not think we need to improve that for now, the error already explain the full context.

## Typescript


<details>
<summary>Code</summary>

```typescript
import { connect } from '@dagger.io/dagger';

connect(
	async (client) => {
		const _ = await client.container().withMountedFile("/test", null).sync()
	},
	{ LogOutput: process.stderr })
```

</details>

<details>
<summary>Output</summary>

```shell
┃ node:internal/process/esm_loader:46                                                                                                                                                 
┃       internalBinding('errors').triggerUncaughtException(                                                                                                                           
┃                                 ^                                                                                                                                                   
┃ [Object: null prototype] {                                                                                                                                                          
┃   [Symbol(nodejs.util.inspect.custom)]: [Function: [nodejs.util.inspect.custom]]   
```

</details>

It actually does not even compile, so we do not really see any stacktrace.
Note that my IDE is able to hint me that the type `null` cannot be used as file and since we do not have pointer in Typescript, we do not really hit this issue :)

![typescript-invalid-type](https://github.com/dagger/dagger/assets/61683879/80a4e9b0-cf18-40d2-987c-54154cde3419)
